File: leetcode/LC1/array/0977_square_of_a_sorted_array.cpp
#include <climits>
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <unordered_map>
using namespace std;

// ex: [-150, -120, -90, 0, 10, 100] use two pointers to left most and right most and find maximum absolute
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        int i = 0;
        int j = nums.size() - 1;
        vector<int> v;
        while(i <= j){
            if(abs(nums[i]) > abs(nums[j])){
                v.push_back(nums[i]*nums[i]);
                i++;
            }else{
                v.push_back(nums[j]*nums[j]);
                j--;
            }
        }
        reverse(v.begin(), v.end());
        return v;
    }
    
};

File: leetcode/LC/array/0977_square_of_a_sorted_array.py
from typing import List

# O(n) | O(1) 
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        start = 0
        end = len(nums) - 1
        
        ans = []
        while(start <= end):
            if abs(nums[start]) > abs(nums[end]):
               ans.append(nums[start]**2)
               start += 1
            else:
               ans.append(nums[end]**2)
               end -= 1
        ans.reverse() # reverses in place, O(n)
        return ans
        
        